src/index.ts

```diff
@@ -47,52 +46,44 @@ interface ConnectSocketOptions {
 
 // Connect an http socket to another tcp server.
 // Based on tunneling proxy code from https://nodejs.org/api/http.html
-const connectSocket = ({
+const connectSocket = async ({
 	cltSocket,
 	hostname,
 	port,
 	head,
 	connect,
 	req,
-}: ConnectSocketOptions) =>
-	connect(
-		port,
-		hostname,
-		cltSocket,
-		req,
-	)
-		.then(srvSocket => {
-			cltSocket.write(
-				`HTTP/1.0 200 Connection Established\r\nProxy-agent: balena-io/node-tunnel (v${version})\r\n\r\n`,
-			);
-			srvSocket.write(head);
-			srvSocket.pipe(
-				cltSocket,
-				{ end: false },
-			);
-			cltSocket.pipe(
-				srvSocket,
-				{ end: false },
-			);
-
-			return Promise.fromCallback(cb => {
-				cltSocket.on('error', cb);
-				srvSocket.on('error', cb);
-				cltSocket.on('end', cb);
-				srvSocket.on('end', cb);
-			}).finally(() => {
-				srvSocket.destroy();
-				cltSocket.destroy();
+}: ConnectSocketOptions) => {
+	try {
+		const srvSocket = await connect(port, hostname, cltSocket, req);
+		cltSocket.write(
+			`HTTP/1.0 200 Connection Established\r\nProxy-agent: balena-io/node-tunnel (v${version})\r\n\r\n`,
+		);
+		srvSocket.write(head);
+		srvSocket.pipe(cltSocket, { end: false });
+		cltSocket.pipe(srvSocket, { end: false });
+
+		try {
+			await new Promise((resolve, reject) => {
+				cltSocket.on('error', err => reject(err));
```

Comment:
This should be able to be:
```suggestion
				cltSocket.on('error', reject);
```